### PR TITLE
issue #3599 fhir-flow better support non IBM FHIR Server upstream systems

### PR DESCRIPTION
**Pull request by punktilious**

### CODE REVIEW

File: fhir-flow/src/main/java/com/ibm/fhir/flow/api/FlowInteraction.java [hunk 1]
```diff
@@ -26,8 +26,8 @@ public abstract class FlowInteraction {
      * @param trackerTicket
      * @param identifier
      */
-    protected FlowInteraction(long changeId, ITrackerTicket trackerTicket, ResourceIdentifier identifier) {
-        this.changeId = changeId;
+    protected FlowInteraction(String entryId, ITrackerTicket trackerTicket, ResourceIdentifier identifier) {
```
Comment (lmsurpre) [inline fhir-flow/src/main/java/com/ibm/fhir/flow/api/FlowInteraction.java:29]: update javadoc to match

File: fhir-flow/src/main/java/com/ibm/fhir/flow/api/IFlowInteractionHandler.java [hunk 2]
```diff
@@ -15,17 +15,17 @@ public interface IFlowInteractionHandler {
...
     /**
      * Perform a create-or-update (PUT) interaction
-     * @param changeId
+     * @param entryId
      * @param identifier
      * @param resourceData
      * @param resource
      */
-    void createOrUpdate(long changeId, ResourceIdentifier identifier, String resourceData, Resource resource);
+    void createOrUpdate(String entryId, ResourceIdentifier identifier, String resourceData, Resource resource);
```
Comment (lmsurpre) [inline fhir-flow/src/main/java/com/ibm/fhir/flow/api/IFlowInteractionHandler.java:30]: will it work if entryId is null?

File: fhir-flow/src/main/java/com/ibm/fhir/flow/api/IFlowInteractionHandler.java [hunk 3]
```diff
@@ -15,17 +15,17 @@ public interface IFlowInteractionHandler {
 
     /**
      * Perform a logical DELETE interaction
-     * @param changeId
+     * @param entryId
      * @param identifier
      */
-    void delete(long changeId, ResourceIdentifier identifier);
+    void delete(String entryId, ResourceIdentifier identifier);
```
Comment (lmsurpre) [inline fhir-flow/src/main/java/com/ibm/fhir/flow/api/IFlowInteractionHandler.java:21]: will it work if entryId is null?

File: fhir-flow/src/main/java/com/ibm/fhir/flow/app/Main.java [hunk 4]
```diff
@@ -132,11 +138,11 @@ public void parseArgs(String[] args) {
...
                 }
                 break;
-            case "--change-id-marker":
+            case "--from-checkpoint":
                 if (i < args.length + 1) {
-                    this.changeIdMarker = Long.parseLong(args[++i]);
+                    this.startFromCheckpoint = args[++i];
                 } else {
-                    throw new IllegalArgumentException("missing value for --change-id-marker");
+                    throw new IllegalArgumentException("missing value for --checkpoint");
```
Comment (lmsurpre) [inline fhir-flow/src/main/java/com/ibm/fhir/flow/app/Main.java:145]:
```suggestion
                    throw new IllegalArgumentException("missing value for --from-checkpoint");
```

File: fhir-flow/src/main/java/com/ibm/fhir/flow/impl/UpstreamFHIRHistoryReader.java [hunk 5]
```diff
@@ -236,56 +294,49 @@ private void processBundle(Bundle bundle) {
...
-            ITrackerTicket trackerTicket = tickets.get(index++);
             // The entry.id field is database resourceId which is used as the PK for the
             // changes table. It just needs to be unique and match the change sequence. It
             // does not need to be contiguous
-            long changeId = Long.parseLong(entry.getId());
             switch (entry.getRequest().getMethod().getValueAsEnum()) {
             case DELETE:
                 // no value to fetch, so just submit the interaction
-                flowWriter.submit(new Delete(changeId, trackerTicket, ResourceIdentifier.from(entry.getFullUrl())));
+                flowWriter.submit(new Delete(entry.getId(), ticket, ResourceIdentifier.from(entry.getFullUrl())));
```
Comment (lmsurpre) [inline fhir-flow/src/main/java/com/ibm/fhir/flow/impl/UpstreamFHIRHistoryReader.java:308]: what happens if entry.getId() is null?

File: fhir-flow/src/main/java/com/ibm/fhir/flow/impl/UpstreamFHIRHistoryReader.java [hunk 6]
```diff
@@ -236,56 +294,49 @@ private void processBundle(Bundle bundle) {
...
+                Resource r = entry.getResource();
+                if (r != null) {
+                    // If the entry already contains the resource, we don't need to issue a new VREAD
+                    // We expect the returned resource to be valid, so it should have an id and meta
+                    ResourceIdentifierVersion riv = new ResourceIdentifierVersion(r.getClass().getSimpleName(), r.getId(), r.getMeta().getVersionId().getValue());
+                    FlowFetchResult ffr = new FlowFetchResult(riv);
+                    ffr.setResource(r);
+                    ffr.setStatus(200); // we didn't have to fetch it, so we're calling it OK
+                    CompletableFuture<FlowFetchResult> flowFetchFuture = CompletableFuture.completedFuture(ffr);
+                    flowWriter.submit(new CreateOrUpdate(entry.getId(), ticket, riv, flowFetchFuture));
```
Comment (lmsurpre) [inline fhir-flow/src/main/java/com/ibm/fhir/flow/impl/UpstreamFHIRHistoryReader.java:321]: what happens if entry.getId() is null?

File: fhir-flow/src/main/java/com/ibm/fhir/flow/api/IFlowInteractionHandler.java (hunk 2 above)
Comment (punktilious) [reply fhir-flow/src/main/java/com/ibm/fhir/flow/api/IFlowInteractionHandler.java:30]: I've added handling in the Bundle processing so we create our own value if the entry.id value is null. Only used for logging to assist in correlating messages.

File: fhir-flow/src/main/java/com/ibm/fhir/flow/api/IFlowInteractionHandler.java (hunk 3 above)
Comment (punktilious) [reply fhir-flow/src/main/java/com/ibm/fhir/flow/api/IFlowInteractionHandler.java:21]: See prior

File: fhir-flow/src/main/java/com/ibm/fhir/flow/app/Main.java (hunk 4 above)
Comment (punktilious) [reply fhir-flow/src/main/java/com/ibm/fhir/flow/app/Main.java:145]: yeah, saw that too when I looked at the diffs! fixed

File: fhir-flow/src/main/java/com/ibm/fhir/flow/impl/UpstreamFHIRHistoryReader.java (hunk 5 above)
Comment (punktilious) [reply fhir-flow/src/main/java/com/ibm/fhir/flow/impl/UpstreamFHIRHistoryReader.java:308]: addressed

File: fhir-flow/src/main/java/com/ibm/fhir/flow/impl/UpstreamFHIRHistoryReader.java (hunk 6 above)
Comment (punktilious) [reply fhir-flow/src/main/java/com/ibm/fhir/flow/impl/UpstreamFHIRHistoryReader.java:321]: addressed

**Review by lmsurpre — APPROVED** [review]
LGTM

**Pull request merged**